src/Renci.SshNet/SshCommand.cs

```diff
@@ -348,21 +367,109 @@ public string EndExecute(IAsyncResult asyncResult)
             {
-                // TODO: check with Oleg if we shouldn't dispose the channel and uninitialize it ?
-                _channel.Dispose();
+                throw new ArgumentNullException(nameof(asyncResult));
             }
+
+            var commandAsyncResult = asyncResult switch
+                {
+                    CommandAsyncResult result when result == _asyncResult => result,
+                    _ => throw new ArgumentException(
+                             $"The {nameof(IAsyncResult)} object was not returned from the corresponding asynchronous method on this class.")
+                };
+
+            lock (_endExecuteLock)
+            {
+                if (commandAsyncResult.EndCalled)
+                {
+                    throw new ArgumentException("EndExecute can only be called once for each asynchronous operation.");
+                }
+
+                // wait for operation to complete (or time out)
+                WaitOnHandle(_asyncResult.AsyncWaitHandle);
+                UnsubscribeFromEventsAndDisposeChannel(_channel);
+                _channel = null;
+
+                commandAsyncResult.EndCalled = true;
+
+                return ExitStatus;
+            }
+        }
+
+        /// <summary>
+        /// Executes the the command asynchronously.
+        /// </summary>
+        /// <returns>Exit status of the operation.</returns>
+        public Task<int> ExecuteAsync()
+        {
+            return ExecuteAsync(forceKill: false, default);
+        }
+
+        /// <summary>
+        /// Executes the the command asynchronously.
+        /// </summary>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <returns>Exit status of the operation.</returns>
+        public Task<int> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            return ExecuteAsync(forceKill: false, cancellationToken);
+        }
+
+        /// <summary>
+        /// Executes the the command asynchronously.
+        /// </summary>
+        /// <param name="forceKill">if true send SIGKILL instead of SIGTERM to cancel the command.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to observe.</param>
+        /// <returns>Exit status of the operation.</returns>
+        public async Task<int> ExecuteAsync(bool forceKill, CancellationToken cancellationToken)
+        {
+#if NET || NETSTANDARD2_1_OR_GREATER
+            await using var ctr = cancellationToken.Register(() => CancelAsync(forceKill), useSynchronizationContext: false).ConfigureAwait(continueOnCapturedContext: false);
+#else
+            using var ctr = cancellationToken.Register(() => CancelAsync(forceKill), useSynchronizationContext: false);
+#endif // NET || NETSTANDARD2_1_OR_GREATER
+
+            try
+            {
+                var status = await Task<int>.Factory.FromAsync(BeginExecute(), EndExecuteWithStatus).ConfigureAwait(false);
```

Comment:
I want to add `Async` overloaded methods but I think this is not a real `async` code. I've prepared a POC here: https://github.com/sshnet/SSH.NET/pull/1318 

I know that your approach is easier to implement, but this is similar to nuget: https://www.nuget.org/packages/Renci.SshNet.Async#versions-body-tab

You can find code here: https://github.com/JohnTheGr8/Renci.SshNet.Async/blob/master/Renci.SshNet.Async/SshNetExtensions.cs